Add config option to disable body or form encoding

shakeelmohamed opened the pull request:
This allows for sending data exactly as it's intended, like cURL. We need this (or a comparable workaround) to POST data without it being altered by `rfc3986()`. See https://github.com/splunk/splunk-javascript-logging/issues/9

I'm happy to add some tests with a bit of guidance!
I saw the following test error on Node 4, but not on 6:

```
/Users/me/git/request/tests/test-headers.js:172
    t.equal(err.message, 'The header content contains invalid characters')
```

mscdex commented:
This option shouldn't be necessary. Why use `.form()`/`form: ...` at all in your case? Just set `body: ...` or pipe the body, with the appropriate `Content-Type` header set.

shakeelmohamed commented:
@mscdex We're sending concatenated JSON blobs, so just setting `json: true` isn't enough. Currently we also have to set the `content-type` header to `application/x-www-form-urlencoded` to avoid passing through `safeStringify()` since that will escape the JSON quotes - which leads to invalid JSON.

Is there another way to to automatically parse JSON responses other than setting `json: true`? We'd like to contain that step in request if at all possible.

chovy commented:
`json: true` has no impact on a request, only the response. Its really just an override if the server doesn't return a proper json header.

simov commented:
@shakeelmohamed a simple code example to reproduce the issue would be great!

shakeelmohamed commented:
Thanks @simov, the following should help - although it's a bit verbose.

Using [httpbin](http://httpbin.org/), I've added example POSTs without and with the `application/x-www-form-urlencoded` header.
I've also added the same examples using request with the latest from npm, and the output with my changes (code snippet at the end).

## curl

### JSON

**This is what we're trying to achieve**.

Note `data` in the response is not encoded, the JSON quotes are just escaped.

```shell
$ curl -v -k http://httpbin.org/post \                                                                                  
    -H "content-type: application/json" \ 
    -d '{"event": "hello", "source":"JavaScript (node.js)"}{"event": "hello2", "source": "JavaScript (node.js)"}'
*   Trying 23.22.14.18...
* Connected to httpbin.org (23.22.14.18) port 80 (#0)
> POST /post HTTP/1.1
> Host: httpbin.org
> User-Agent: curl/7.43.0
> Accept: */*
> content-type: application/json
> Content-Length: 104
> 
* upload completely sent off: 104 out of 104 bytes
< HTTP/1.1 200 OK
< Server: nginx
< Date: Wed, 19 Oct 2016 23:17:43 GMT
< Content-Type: application/json
< Content-Length: 441
< Connection: keep-alive
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< 
{
  "args": {}, 
  "data": "{\"event\": \"hello\", \"source\":\"JavaScript (node.js)\"}{\"event\": \"hello2\", \"source\": \"JavaScript (node.js)\"}", 
  "files": {}, 
  "form": {}, 
  "headers": {
    "Accept": "*/*", 
    "Content-Length": "104", 
    "Content-Type": "application/json", 
    "Host": "httpbin.org", 
    "User-Agent": "curl/7.43.0"
  }, 
  "json": null, 
  "origin": "<snip>", 
  "url": "http://httpbin.org/post"
}
* Connection #0 to host httpbin.org left intact
```

### urlencoded

```shell
$ curl -v -k http://httpbin.org/post \
    -H "content-type: application/x-www-form-urlencoded" \
    -d '{"event": "hello", "source":"JavaScript (node.js)"}{"event": "hello2", "source": "JavaScript (node.js)"}'
*   Trying 23.22.14.18...
* Connected to httpbin.org (23.22.14.18) port 80 (#0)
> POST /post HTTP/1.1
> Host: httpbin.org
> User-Agent: curl/7.43.0
> Accept: */*
> content-type: application/x-www-form-urlencoded
> Content-Length: 104
> 
* upload completely sent off: 104 out of 104 bytes
< HTTP/1.1 200 OK
< Server: nginx
< Date: Wed, 19 Oct 2016 23:18:11 GMT
< Content-Type: application/json
< Content-Length: 472
< Connection: keep-alive
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< 
{
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {
    "{\"event\": \"hello\", \"source\":\"JavaScript (node.js)\"}{\"event\": \"hello2\", \"source\": \"JavaScript (node.js)\"}": ""
  }, 
  "headers": {
    "Accept": "*/*", 
    "Content-Length": "104", 
    "Content-Type": "application/x-www-form-urlencoded", 
    "Host": "httpbin.org", 
    "User-Agent": "curl/7.43.0"
  }, 
  "json": null, 
  "origin": "<snip>", 
  "url": "http://httpbin.org/post"
}
* Connection #0 to host httpbin.org left intact
```

## using request library

In the urlencoded request, we have `form: { '{"event":"hello","source":"JavaScript (node.js)"}{"event":"hello","source":"JavaScript (node.js)"}': '' }`.
This isn't evident in the response from httpbin, but **the parentheses (for example) get encoded in the POST body into `%28` and `%29` before the request is made**.

### JSON

```shell
$ node index.js            
REQUEST { url: 'http://httpbin.org/post',
  body: '{"event":"hello","source":"JavaScript (node.js)"}{"event":"hello","source":"JavaScript (node.js)"}',
  json: true,
  callback: [Function],
  method: 'POST' }
REQUEST make request http://httpbin.org/post
REQUEST onRequestResponse http://httpbin.org/post 200 { server: 'nginx',
  date: 'Wed, 19 Oct 2016 23:48:17 GMT',
  'content-type': 'application/json',
  'content-length': '562',
  connection: 'close',
  'access-control-allow-origin': '*',
  'access-control-allow-credentials': 'true' }
REQUEST reading response's body
REQUEST finish init function http://httpbin.org/post
REQUEST response end http://httpbin.org/post 200 { server: 'nginx',
  date: 'Wed, 19 Oct 2016 23:48:17 GMT',
  'content-type': 'application/json',
  'content-length': '562',
  connection: 'close',
  'access-control-allow-origin': '*',
  'access-control-allow-credentials': 'true' }
REQUEST end event http://httpbin.org/post
REQUEST has body http://httpbin.org/post 562
REQUEST emitting complete http://httpbin.org/post
***non-urlencoded:***
{ args: {},
  data: '"{\\"event\\":\\"hello\\",\\"source\\":\\"JavaScript (node.js)\\"}{\\"event\\":\\"hello\\",\\"source\\":\\"JavaScript (node.js)\\"}"',
  files: {},
  form: {},
  headers: 
   { Accept: 'application/json',
     'Content-Length': '116',
     'Content-Type': 'application/json',
     Host: 'httpbin.org' },
  json: '{"event":"hello","source":"JavaScript (node.js)"}{"event":"hello","source":"JavaScript (node.js)"}',
  origin: '<snip>',
  url: 'http://httpbin.org/post' }
```

### urlencoded

```shell
$ node index.js
REQUEST { url: 'http://httpbin.org/post',
  body: '{"event":"hello","source":"JavaScript (node.js)"}{"event":"hello","source":"JavaScript (node.js)"}',
  json: true,
  headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
  callback: [Function],
  method: 'POST' }
REQUEST make request http://httpbin.org/post
REQUEST onRequestResponse http://httpbin.org/post 200 { server: 'nginx',
  date: 'Wed, 19 Oct 2016 23:47:41 GMT',
  'content-type': 'application/json',
  'content-length': '445',
  connection: 'close',
  'access-control-allow-origin': '*',
  'access-control-allow-credentials': 'true' }
REQUEST reading response's body
REQUEST finish init function http://httpbin.org/post
REQUEST response end http://httpbin.org/post 200 { server: 'nginx',
  date: 'Wed, 19 Oct 2016 23:47:41 GMT',
  'content-type': 'application/json',
  'content-length': '445',
  connection: 'close',
  'access-control-allow-origin': '*',
  'access-control-allow-credentials': 'true' }
REQUEST end event http://httpbin.org/post
REQUEST has body http://httpbin.org/post 445
REQUEST emitting complete http://httpbin.org/post
***non-urlencoded:***
{ args: {},
  data: '',
  files: {},
  form: { '{"event":"hello","source":"JavaScript (node.js)"}{"event":"hello","source":"JavaScript (node.js)"}': '' },
  headers: 
   { Accept: 'application/json',
     'Content-Length': '106',
     'Content-Type': 'application/x-www-form-urlencoded',
     Host: 'httpbin.org' },
  json: null,
  origin: '<snip>',
  url: 'http://httpbin.org/post' }
```

### With my changes

**Note that nothing is encoded here, and the full body isn't quoted.**

```shell
$ node index.js
REQUEST { url: 'http://httpbin.org/post',
  body: '{"event":"hello","source":"JavaScript (node.js)"}{"event":"hello","source":"JavaScript (node.js)"}',
  json: true,
  callback: [Function],
  method: 'POST' }
REQUEST make request http://httpbin.org/post
REQUEST onRequestResponse http://httpbin.org/post 200 { server: 'nginx',
  date: 'Wed, 19 Oct 2016 23:50:59 GMT',
  'content-type': 'application/json',
  'content-length': '413',
  connection: 'close',
  'access-control-allow-origin': '*',
  'access-control-allow-credentials': 'true' }
REQUEST reading response's body
REQUEST finish init function http://httpbin.org/post
REQUEST response end http://httpbin.org/post 200 { server: 'nginx',
  date: 'Wed, 19 Oct 2016 23:50:59 GMT',
  'content-type': 'application/json',
  'content-length': '413',
  connection: 'close',
  'access-control-allow-origin': '*',
  'access-control-allow-credentials': 'true' }
REQUEST end event http://httpbin.org/post
REQUEST has body http://httpbin.org/post 413
REQUEST emitting complete http://httpbin.org/post
***non-urlencoded:***
{ args: {},
  data: '{"event":"hello","source":"JavaScript (node.js)"}{"event":"hello","source":"JavaScript (node.js)"}',
  files: {},
  form: {},
  headers: 
   { Accept: 'application/json',
     'Content-Length': '98',
     'Content-Type': 'application/json',
     Host: 'httpbin.org' },
  json: null,
  origin: '<snip>',
  url: 'http://httpbin.org/post' }
```

### code

```javascript
var request = require("request");

var event = {
    event: "hello",
    source: "JavaScript (node.js)"
};

var body = JSON.stringify(event) + JSON.stringify(event);

var options = {
    url: "http://httpbin.org/post",
    body: body,
    json: true
};

options.headers = {
    "Content-Type": "application/x-www-form-urlencoded"
};

request.post(options, function(err, resp, body) {
    console.log("***non-urlencoded:***");
    console.log(body);
});

```

chovy commented:
I'm unclear what you're trying to achieve. It seems you're trying to work with a broken API. Are you unable to make your API work properly?

`body: ...` sends a JSON payload and `form: ...` sends a URL encoded body. I've never come across any need for anything other than those two in all my years of scraping and using APIs with request.

shakeelmohamed commented:
> It seems you're trying to work with a broken API. Are you unable to make your API work properly?

The API works this way by design, and this minimal change addresses our problem.

If this PR is totally unreasonable (I don't think it is), then we'll use something else.

simov commented:
@shakeelmohamed can you run this code:

```js
var server = http.createServer()
server.on('request', (req, res) => {
  var data = ''
  req.on('data', (chunk) => {
    data += chunk
  })
  req.on('end', () => {
    console.log(data)
    res.end()
  })
})
server.listen(3000)

var event = {
  event: 'hello',
  source: 'JavaScript (node.js)'
}

request({
  method: 'POST',
  url: 'http://localhost:3000',
  headers: {
    'Content-Type': 'application/x-www-form-urlencoded'
  },
  body: JSON.stringify(event) + JSON.stringify(event),
  callback: (err, res, body) => {
    console.log('DONE!')
  }
})
```

Is this the request body you are aiming for:

```
{"event":"hello","source":"JavaScript (node.js)"}{"event":"hello","source":"JavaScript (node.js)"}
```

If that's the case you'll have to `JSON.parse` the response body manually.

shakeelmohamed commented:
Thanks @simov. The workaround is now, setting `json: false` and manually parsing the JSON response.

That said, is this PR moot?

chovy commented:
i don't understand why you're posting stringified json as form data. why not just send it as a json body?

simov commented:
I agree with @chovy that the use case seems a bit weird. Sending content type `urlencoded` with stringified JSON data is just wrong.

@shakeelmohamed rfc3986 is used strictly for `urlencoded` data and if your data was in fact `urlencoded` you would be able to pass the `encode: false` and/or `format : 'RFC1738'` option to the underlying `qs` module.

shakeelmohamed commented:
Alright, thanks everyone. I have a workaround so I'll close this.
